### PR TITLE
GatherNDGrad for CPU

### PR DESCRIPTION
**Pull request by pengwa**
**Description**: CPU kernel implementation for GatherNDGrad
**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

File: onnxruntime/core/providers/cpu/tensor/gather_nd.cc [hunk 1]
```diff
@@ -125,15 +102,55 @@ Status GatherNDBase::PrepareForCompute(OpKernelContext* context, Prepare& p, con
                         : ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "invalid index found, index = ", err_index);
 }
 
-template Status GatherNDBase::PrepareForCompute<int32_t>(OpKernelContext*, Prepare&, concurrency::ThreadPool*) const;
-template Status GatherNDBase::PrepareForCompute<int64_t>(OpKernelContext*, Prepare&, concurrency::ThreadPool*) const;
+template Status GatherNDBase::PrepareForCompute<int32_t>(const TensorShape& input_shape,
+                                                         const Tensor* indices_tensor,
+                                                         const int64_t bytes_per_value, Prepare& p,
+                                                         concurrency::ThreadPool*) const;
```
Comment (ytaous) [inline onnxruntime/core/providers/cpu/tensor/gather_nd.cc:108]: inconsistent signature -either with or without param names - same for the one below

File: orttraining/orttraining/training_ops/cpu/tensor/gather_nd_grad.cc [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+                                                     DataTypeImpl::GetTensorType<int32_t>()}),
+                        GatherNDGrad);
+
+}  // namespace contrib
+
+#endif
+
+template <typename InputT>
+struct GatherNDGradComputeImpl {
+  void operator()(GatherNDBase::Prepare& p, const Tensor* update_tensor) const {
```
Comment (ytaous) [inline orttraining/orttraining/training_ops/cpu/tensor/gather_nd_grad.cc:26]: on windows build - error C2248: 'onnxruntime::GatherNDBase::Prepare':
cannot access protected struct declared in class 'onnxruntime::GatherNDBase'

File: orttraining/orttraining/test/training_ops/cpu/tensor/gather_nd_grad_op_test.cc [hunk 3]
```diff
@@ -74,7 +86,30 @@ TEST(GatherNDGradOpTest, GatherNDGrad_batch_dims_of_2) {
...
+  OpTester test("GatherNDGrad", 1, kMSDomain);
+  test.AddAttribute<int64_t>("batch_dims", 2);
+  test.AddInput<int64_t>("shape", {4}, {2, 2, 2, 3});
+  test.AddInput<int64_t>(
+      "indices", {2, 2, 1},
+      {
+          -1,  // batch 0
+          -1,  // batch 1
+          -2,  // batch 2
+          1,   // batch 3
```
Comment (ytaous) [inline orttraining/orttraining/test/training_ops/cpu/tensor/gather_nd_grad_op_test.cc:100]: nit - would it look better to shift right by 1 space?

File: orttraining/orttraining/test/training_ops/cpu/tensor/gather_nd_grad_op_test.cc [hunk 4]
```diff
@@ -20,8 +19,20 @@ TEST(GatherNDGradOpTest, GatherNDGrad_slice_float_int64_t_batch_dims_1) {
...
+  test.AddInput<int64_t>("shape", {3}, {2LL, 2LL, 3LL});
+  test.AddInput<int64_t>("indices", {2, 2}, {-2LL, -1LL, -1LL, -2LL});
+  test.AddInput<float>("update", {2, 3}, ValueRange(6, 1.0f));
+  test.AddOutput<float>("output", {2, 2, 3}, {0, 0, 0, 1, 2, 3, 4, 5, 6, 0, 0, 0});
+  test.Run();
+}
+
 TEST(GatherNDGradOpTest, GatherNDGrad_slice_double_int32_t_batch_dims_3) {
-  if (!HasCudaEnvironment(600 /*min_cuda_architecture*/)) return;
+  if (NeedSkipIfCudaArchLowerThan(600)) {
```
Comment (ytaous) [inline orttraining/orttraining/test/training_ops/cpu/tensor/gather_nd_grad_op_test.cc:33]: I remember "#if __CUDA_ARCH__ >= xxx" won't work properly in .cc. since you are on it, although it's out of scope, would it be ok that we also clean up the code in gather_grad_op_test.cc and gather_grad.cc using the same method? Or we can do this in separate PR. @weixingzhang

Comment (pengwa) [reply orttraining/orttraining/test/training_ops/cpu/tensor/gather_nd_grad_op_test.cc:33]: ya, make sense! yes let me see whether I can fix them...

File: orttraining/orttraining/test/training_ops/cpu/tensor/gather_nd_grad_op_test.cc (hunk 3 above)
Comment (pengwa) [reply orttraining/orttraining/test/training_ops/cpu/tensor/gather_nd_grad_op_test.cc:100]: I think this is already aligned, considering there is negative symbol in line 99.  Correct me if I misundertood you. Thank you! :)

File: orttraining/orttraining/training_ops/cpu/tensor/gather_nd_grad.cc (hunk 2 above)
Comment (pengwa) [reply orttraining/orttraining/training_ops/cpu/tensor/gather_nd_grad.cc:26]: thanks!, will fix it!!

File: orttraining/orttraining/training_ops/cuda/tensor/gather_grad.cc [hunk 5]
```diff
@@ -76,13 +76,10 @@ Status DispatchToGatherGradImpl(
   if (utils::IsPrimitiveDataType<float>(t_data_type)) {
     return DispatchToGatherGradImplByTin<float>(
         tin_data_type, cuda_kernel, num_weights, stride, num_inputs, param_itrs, grad, indices, output);
-  }
-#if !defined(__CUDA_ARCH__) || __CUDA_ARCH__ >= 700
-  else if (utils::IsPrimitiveDataType<MLFloat16>(t_data_type)) {
+  } else if (utils::IsPrimitiveDataType<MLFloat16>(t_data_type)) {
```
Comment (ytaous) [inline orttraining/orttraining/training_ops/cuda/tensor/gather_grad.cc:79]: is this ok to simply remove the check?

Comment (pengwa) [reply orttraining/orttraining/training_ops/cuda/tensor/gather_grad.cc:79]: I think tis' okay. If the GPU is not supposed to support MLFloat16. it will fail anyway, I checked Gemm, Matmul, they don't check the hardware when calling into MLFloat16 code path, am I missing something?

Comment (ytaous) [reply orttraining/orttraining/training_ops/cuda/tensor/gather_grad.cc:79]: I wonder why it 's there initially.

Comment (ytaous) [reply orttraining/orttraining/training_ops/cuda/tensor/gather_grad.cc:79]: @edgchen1 - can you please review this part? Looks like you refactored this code a while ago.

Comment (edgchen1) [reply orttraining/orttraining/training_ops/cuda/tensor/gather_grad.cc:79]: i think so too. `__CUDA_ARCH__` is not supposed to be used in host code.
https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html#virtual-architecture-identification-macro
i probably copied it over from somewhere :)

File: orttraining/orttraining/training_ops/cpu/tensor/gather_nd_grad.cc [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    const int64_t grad_size = update_tensor->Shape().Size();
+    const int64_t slice_size = p.element_count_per_slice;
+    const InputT* input_base_casted = reinterpret_cast<const InputT*>(p.input_base);
+    InputT* output_base_casted = reinterpret_cast<InputT*>(p.output_base);
+#ifdef _OPENMP
+#pragma omp parallel for
```
Comment (edgchen1) [inline orttraining/orttraining/training_ops/cpu/tensor/gather_nd_grad.cc:32]: i think we should not use #pragma omp directly, see this doc:
https://github.com/microsoft/onnxruntime/blob/master/docs/NotesOnThreading.md

File: orttraining/orttraining/training_ops/cuda/tensor/gather_grad.cc (hunk 5 above)
Comment (ytaous) [reply orttraining/orttraining/training_ops/cuda/tensor/gather_grad.cc:79]: should we replace the deleted check with NeedSkipIfCudaArchLowerThan() and/or HasCudaEnvironment() ?

Comment (edgchen1) [reply orttraining/orttraining/training_ops/cuda/tensor/gather_grad.cc:79]: I think the condition was always true before. The half versions of GatherGradImpl() are instantiated without any such check in gather_grad_impl.cu. So I think we don't need a check here.

File: orttraining/orttraining/training_ops/cpu/tensor/gather_nd_grad.cc (hunk 6 above)
Comment (pengwa) [reply orttraining/orttraining/training_ops/cpu/tensor/gather_nd_grad.cc:32]: oh yeah, good point.

**Comment by pengwa** [comment]
thank you @ytaous !
